lib/utils/ttyname.go

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+// #include <unistd.h>
+import "C"
+import (
+	"errors"
+	"os"
+)
+
+// TtyName fetches the OS TTY name for a file descriptor.
+// This returns an error if the file descriptor is not associated with a TTY.
+func TtyName(file *os.File) (*string, error) {
+	fd := file.Fd()
+	ttyNameBuf := [256]C.char{}
+	status := C.ttyname_r((C.int)(fd), &ttyNameBuf[0], 256)
```

Comment:
I think this should be possible to implement natively w/o cgo. Also, dependency on `unistd` makes this linux-specific and won't compile on darwin I assume?

Reply:
@xacrimon I think you should just be able to do `os.Readlink("/proc/self/fd/0")` in the `reexec.go` file to get this, take a look:

```
$ pwd
/proc/self/fd
$ ll
total 0
dr-x------ 2 rjones rjones  0 Feb  5 17:13 ./
dr-xr-xr-x 9 rjones rjones  0 Feb  5 17:13 ../
lrwx------ 1 rjones rjones 64 Feb  5 17:13 0 -> /dev/pts/0
lrwx------ 1 rjones rjones 64 Feb  5 17:13 1 -> /dev/pts/0
lrwx------ 1 rjones rjones 64 Feb  5 17:13 2 -> /dev/pts/0
lrwx------ 1 rjones rjones 64 Feb  6 01:11 255 -> /dev/pts/0
$ readlink -f 0
/dev/pts/0
```

Reply:
I've done some research and I don't think this is possible to implement without cgo. @russjones great thought but that only allows you to resolve file descriptors for the current processes stdout/stderr/stdin binding, i.e the one allocated by your terminal emulator. We need to do this for arbitrary TTYs.

Reply:
Also MacOS implements the same header and supplies the same `ttyname(1)` function so it'll work fine there.

Reply:
Please chime in if you know of a better way but I haven't found one.

Reply:
That said you need to have Xcode installed to supply the headers.

Reply:
Missed your other comment. Yeah avoiding it looks doable.